removing max7456 dependencies from osd code (fixed positions etc)

fishpepper opened the pull request:
I started to work on removing max7456 dependencies from the osd code.
This is work in progress, do not merge it yet ;)
I just post it at this stage to get early feedback.

This prepares for tinyOSD integration. See issue #3397 

Done so far:
- renamed rows+cols to rowCount/colCount
- reworked the position bit stuffing from osdConfig_s struct (use 2x8bit + 1 visibility bool)

todo: 
- make default positions (e.g. center positions) dependent on rowCount+colCount
- change msp bit stuffing for osdconfig data (allows only 31 char width)

DanNixon commented on src/main/io/osd.h:
```diff
@@ -98,7 +98,8 @@ typedef enum {
 typedef struct osdConfig_s {
-    uint16_t item_pos[OSD_ITEM_COUNT];
+    uint8_t item_pos[OSD_ITEM_COUNT][2];
+    bool visible[OSD_ITEM_COUNT];
```
An array of

```
struct {
    uint8_t x;
    uint8_t y;
    bool visible;
}
```

would help keep the default positions tidier (should be more efficient to iterate over too).

ledvinap replied:
@DanNixon: :+1: 
Unless it complicates CLI a lot ...

fishpepper commented:
I was wondering if all entries should become configurable. For example OSD_CROSSHAIRS is hard coded in osd.c:
`
    case OSD_CROSSHAIRS:
        elemPosX = 14 - 1; // Offset for 1 char to the left
        elemPosY = 6;
      ...`
What about having this (and others) configurable as well?

mikeller commented:
Yeah, totally for it. Being able to adjust the crosshairs / artificial horizon to match the horizon at 'cruising speed' makes sense.

hydra commented on src/main/target/KROOZX/config.c:
```diff
@@ -36,6 +36,8 @@ void targetConfiguration(void)
     compassConfigMutable()->mag_hardware = 0;
-    osdConfigMutable()->item_pos[OSD_MAIN_BATT_VOLTAGE] = OSD_POS(12, 1) | VISIBLE_FLAG;
+    osdConfigMutable()->item_pos[OSD_MAIN_BATT_VOLTAGE][0] = 12;
+    osdConfigMutable()->item_pos[OSD_MAIN_BATT_VOLTAGE][1] = 1;
```
0 and 1 are magic numbers, replace with defines for X and Y or ROW/COLUMN as appropriate.

ledvinap replied:
`struct` is even better

ledvinap commented on src/main/drivers/display.h:
```diff
@@ -21,8 +21,8 @@ struct displayPortVTable_s;
-    uint8_t rows;
-    uint8_t cols;
+    uint8_t rowCount;
+    uint8_t colCount;
```
I don't see why should be naming better. rows/cols has very concrete meaning when describing display ...

martinbudden replied:
Renaming was at my suggestion. `Count` is the standard suffix used in betaflight.

ledvinap replied:
I don't see any advantage in this context .. but both will work

ledvinap commented on src/main/fc/fc_msp.c:
```diff
@@ -2171,7 +2176,13 @@ static mspResult_e mspCommonProcessInCommand(uint8_t cmdMSP, sbuf_t *src)
                     /* Set element positions */
-                    osdConfigMutable()->item_pos[addr] = value;
+                    //FIXME: this should be changed to separate x and y transfer?!
+                    uint8_t x = (value & ((1 << 5) - 1));
```
`int x` will be cleaner

ledvinap commented on src/main/io/osd.c:
```diff
@@ -690,46 +690,160 @@ static void osdDrawElements(void)
-    osdConfig->item_pos[OSD_NUMERICAL_VARIO]    = OSD_POS(23, 8)  | VISIBLE_FLAG;
-    osdConfig->item_pos[OSD_ESC_TMP]            = OSD_POS(18, 2)  | VISIBLE_FLAG;
-    osdConfig->item_pos[OSD_ESC_RPM]            = OSD_POS(19, 2)  | VISIBLE_FLAG;
+    osdConfig->item_pos[OSD_RSSI_VALUE][0] = 8;
```
Using macro here will make code implementation-independent:
```
#define OSD(_id, _x, _y) do { osdConfig->item_pos[_id][0] = (_x); \
                              osdConfig->item_pos[_id][1] = (_y); \
                              osdConfig->visible[_id] = true; \
                         } while(0)
/**/
OSD(OSD_RSSI_VALUE, 8, 2);

```
and `#undef` OSD at end of this function

ledvinap commented on src/main/io/osd.h:
```diff
@@ -28,11 +28,11 @@
-#define OSD_POS(x,y)  ((x & OSD_POSITION_XY_MASK) | ((y & OSD_POSITION_XY_MASK) << OSD_POSITION_BITS))
-#define OSD_X(x)      (x & OSD_POSITION_XY_MASK)
-#define OSD_Y(x)      ((x >> OSD_POSITION_BITS) & OSD_POSITION_XY_MASK)
+//#define OSD_POSITION_BITS 5 // 5 bits gives a range 0-31
```
This abstraction is still used when sending OSD position over MSP. But it may be moved into some MSP-related file

ledvinap commented on src/main/fc/fc_msp.c:
```diff
@@ -846,7 +846,12 @@ static bool mspCommonProcessOutCommand(uint8_t cmdMSP, sbuf_t *dst, mspPostProce
         sbufWriteU16(dst, osdConfig()->alt_alarm);
         for (int i = 0; i < OSD_ITEM_COUNT; i++) {
-            sbufWriteU16(dst, osdConfig()->item_pos[i]);
+            //FIXME: this should be changed to separate x, y, and visibility
```
Macros from osd.h ...

fishpepper commented:
I modified the pull request a bit. OSD positions can now be entered relative to 8+1 directions (Center, N, NE, E, SE, S,...).
This way they stay at the intended positions no matter what screen resolution you have.
For the configurator I would suggest storing the position relative to the closest of the 9 origin positions.

How does this sound to you?
I know the build fails currently for some targets (out of mem). I will do a pull request for that soon.

ledvinap commented:
@fishpepper : Is anchoring really useful? Display resolution is fixed for single quad, so placement will be equivalent to absolute for normal user.
Auto-selecting anchor in configurator can have unexpected results - column of NW items will tear from W items.
Reset defaults will benefit from this,  so it may be implemented in resetConfing?

ledvinap commented on src/main/io/osd.c:
```diff
@@ -227,16 +227,64 @@ static uint8_t osdGetDirectionSymbolFromHeading(int heading)
+// to absolute positioning (based on display height and width)
+static void osdConvertToAbsolutePosition(uint8_t item, int8_t *pos_x, int8_t *pos_y) {
+    // output display dimensions
+    uint8_t display_width  = osdDisplayPort->colCount - 1;
```
Why `-1` ?

fishpepper replied:
I moved the -1 from all calculations up to the variable definition. The name _width makes no longer sense, you are right. should be renamed.

ledvinap commented on src/main/io/osd.c:
```diff
@@ -227,16 +227,64 @@ static uint8_t osdGetDirectionSymbolFromHeading(int heading)
+    uint8_t origin = osdConfig()->item_pos[item].origin;
+
+    // start with center
+    elemPosX += display_width / 2;
```
You may start with zero, add `width/2` if not west and another  `width/2` if east

fishpepper replied:
good idea!

ledvinap commented on src/main/io/osd.c:
```diff
@@ -227,16 +227,64 @@ static uint8_t osdGetDirectionSymbolFromHeading(int heading)
+
+
+    // make sure to return valid x/y positions \in [0..max]
+    *pos_x = MAX(0, MIN(display_width, elemPosX));
```
constrain(elemPosX, 0, osdDisplayPort->colCount - 1);

fishpepper replied:
thanks, will use that instead

fishpepper commented:
You are right, resolution will probably not change. However changing cameras with PAL or NTSC output are something I can imagine that people do durng a quads life. Some vtx/cam combos for the micro quads also have a button to change the output format on the fly. With relative positioning the data will always stay at the proper position (borders). This was the reason to have it system wide instead of on reset/initialization.

ledvinap commented on src/main/io/osd.h:
```diff
@@ -97,8 +97,32 @@ typedef enum {
+#define OSD_ORIGIN_MASK_W (1<<3)
+
+typedef enum {
+    OSD_ORIGIN_C  = 0,
```
```
typedef enum {
  OSD_ORIGIN_C = 0,
  OSD_ORIGIN_N = 1 << 0,
..
   OSD_ORIGIN_NW = OSD_ORIGIN_N | OSD_ORIGIN_W,
..

fishpepper replied:
Interesting, I did not expect this to work (but it does). Will use that.

ledvinap commented on src/main/io/osd.h:
```diff
@@ -97,8 +97,32 @@ typedef enum {
+
 typedef struct osdConfig_s {
-    uint8_t item_pos[OSD_ITEM_COUNT][2];
+    osdItemPos_t item_pos[OSD_ITEM_COUNT];
```
`visible` may be moved into osdItem struct ...

fishpepper replied:
Then this will need some renaming like this:
```
typedef struct {
    int8_t pos_x;
    int8_t pos_y;
    osd_origin_e origin;
    bool visible;
} osdItem_t;

typedef struct osdConfig_s {
    osdItem_t item[OSD_ITEM_COUNT];

    bool enabled_stats[OSD_STAT_COUNT];

    // Alarms
    uint8_t rssi_alarm;
    uint16_t cap_alarm;
    uint16_t time_alarm;
    uint16_t alt_alarm;

    osd_unit_e units;
} osdConfig_t;
```

ledvinap replied:
Yes. Personally I don't see any disadvantage, just put all osdItem related config together ...

ledvinap commented on src/main/fc/fc_msp.c:
```diff
@@ -2180,8 +2181,10 @@ static mspResult_e mspCommonProcessInCommand(uint8_t cmdMSP, sbuf_t *src)
-                    osdConfigMutable()->item_pos[addr][0] = x;
-                    osdConfigMutable()->item_pos[addr][1] = y;
+                    uint8_t origin = (value & 0xF000) >> 12;
+                    osdConfigMutable()->item_pos[addr].x = x;
```
With anchoring, item position has to be signed (and 5 bits may be too small unless you choose 'close' reference point)

fishpepper replied:
I know, if the idea of the anchoring is approved I will rewrite this code and the associated code in the configurator gui.

martinbudden commented:
@fishpepper , how close is this to being complete?

I think the changes you have made are worthwhile even if not fully complete, so it would be good to merge your changes.

fishpepper commented:
I got distracted by other work. I have no access to a max7456 based fc so that I can not test the changes right now. I can only test it once I finished the whole tinyOSD integration.

My plan was to address the suggested changes (especially msp) and write a patch for the configurator ui to handle the new msp format in the next days. Probably today in the evening.

fishpepper commented:
I did the rebase but I am running into the next problem :-\

The CLI is assuming one 16 bit value for position. See settings.c:
```{ "osd_vbat_pos",               VAR_UINT16  | MASTER_VALUE, .config.minmax = { 0, OSD_POSCFG_MAX }, PG_OSD_CONFIG, offsetof(osdConfig_t, item[OSD_MAIN_BATT_VOLTAGE]) },```

I will need to change this. My suggestion would be:
``` { "osd_item_vbat",               MODE_ARRAY | MASTER_VALUE, .config.array.length = 3, PG_OSD_CONFIG, offsetof(osdConfig_t, item[OSD_MAIN_BATT_VOLTAGE]) },```

That would require the osdItem_t struct to be 3byte aligned, e.g. 
```
typedef struct {
    int8_t x;
    int8_t y;
    uint8_t flags; // <-- visible and origin will go into this
} osdItem_t;
```
with flags holding the origin and visible flag. Not that nice :-\
Or an unnamed struct/union?

Before I put a lot of effort in it:  
It looks like MODE_ARRAY is not used right now in the cli?! Is this an unmaintained leftover?
Can I revive it? It would be nice to set the osd positions in human readable format from the cli.
Still need to figure out how to handle the signedness then...

martinbudden commented:
I think MODE_ARRAY was recently added to support bind values for SPI connected receivers, but I don't think that has all been implemented yet. I see no reason it shouldn't be used for the OSD settings.

fishpepper commented:
Great!
I had a look at the array implementation and enhanced it a bit and made it a bit more efficient (#3523)

DanNixon commented on src/main/cms/cms_menu_osd.c:
```diff
@@ -40,64 +40,64 @@
+    {"HORIZON",            OME_VISIBLE, NULL, &osdConfig_item[OSD_ARTIFICIAL_HORIZON], 0},
+    {"HORIZON SIDEBARS",   OME_VISIBLE, NULL, &osdConfig_item[OSD_HORIZON_SIDEBARS], 0},
+    {"TIMER 1",            OME_VISIBLE, NULL, &osdConfig_item[OSD_TIMER_1], 0},
+    {"TIMER 2",            OME_VISIBLE, NULL, &osdConfig_item[OSD_TIMER_2], 0},
```
`OSD_ITEM_TIMER_x` is correct here.

hydra replied:
can we use a prefix on all of the 'item' defines to clarify what is and what is not an 'item'.

also, personally i prefer 'element' to 'item'

DanNixon replied:
See #3477

fishpepper commented:
I can not test the positioning changes right now, I am still waiting for the fc with an integrated max7456 to arrive... Once I get it I will verify the changes. 
I can add _ITEM or _ELEMENT to the pull request. I have no preference on which one to use.

DanNixon commented:
@fishpepper If possible could you modify the layout of the flags such that the visibility bit is next to the free bits? A future improvement is to allow different pages by having multiple visibility bits for each element and keeping all visibility bits together would be preferable.

Is all that is required for this to be tested with a MAX7456 based OSD? If so I can have a look at it today.

fishpepper commented:
ok moved the origin bits to the higher 4 bits, bit0 = visible, bit1..11 free for future use

Yes, I can not test it yet with a max7456 based board.
Please note that the configurator will not work as expected, the msp changes are not merged into the configurator ui.

DanNixon commented:
Cool, I'll give it a test in the morning.

Do you have a branch with the required configurator changes?

DanNixon commented:
@fishpepper The changes work well for me, there are a couple of naming issues remaining that I mentioned above.

I've corrected them and added a unit test for the position anchoring, see https://github.com/fishpepper/betaflight/pull/1

martinbudden commented:
@fishpepper , @DanNixon , do you think this can be merged in now?

DanNixon commented:
@martinbudden It needs (at least the fixes from) https://github.com/fishpepper/betaflight/pull/1 and configurator changes (I can probably get that done today).

As it is the MSP changes break backwards compatibility, are we OK with that?

fishpepper commented:
No, do not merge it right now, It breaks the osd configurator.
I will create a pull request for the configurator (probably today in the evening).
API_VERSION_MINOR should be increased in order to detect the new MSP format.

DanNixon commented:
@fishpepper The MSP version is incremented as part of the release.
Checking for a version greater than or equal to 1.36.0 is sufficient.

fishpepper commented:
I started to work on the necessary configurator changes. It is not tested and wip. It took more time than expected as I never coded chrome / js apps before :)
I am running out of time for today, I will continue the next days.
https://github.com/fishpepper/betaflight-configurator/tree/osd_pos_fix

fishpepper commented:
Seems like I understood how the js stuff for the configurator works ;)

@DanNixon: Could you test the code from
https://github.com/fishpepper/betaflight-configurator/tree/osd_pos_fix
with your max7456?

DanNixon commented:
@fishpepper I notice a lot of inconsistencies between what the preview shows and what the OSD actually displays. Sometimes the elements appear two rows out of where they are positioned in the configurator, sometimes they are a column out, some elements are not visible despite being enabled.

I'll have another look in the morning and see if I can find why this is happening.

DanNixon commented:
Ok, my bad. I flashed an incorrect firmware last night.

It is in fact working correctly.
There are just a few default positions that are off, I'll have a look at correcting those now.

fishpepper commented:
I think the relative to absolute conversion is still wrong in bf...
I fixed it in bf-configurator, the preview there is now correct. I will create some unit tests and fix it for bf the next days.

DanNixon commented:
I think my confusion was down to the fact the configurator always shows a PAL preview regardless of the actual video system (which in my case was NTSC).

The reduced number of video lines meant that elements anchored north and centre may overlap of step over each other when changing video systems.

As far as I can tell the firmware is doing what it should and the issue is with the configurator.

DanNixon commented:
With regards to the default positions, this seems to be a reasonable default: https://github.com/DanNixon/betaflight/commit/7f3bdb64b2d92684eb495fa914bc403e1cd8261d

However I feel a better idea is to set the default position of every item to the centre of the screen (i.e. 0, 0 anchored in the centre) and set them all to be invisible.
This way a user can enable only what they want and position them as they want and it gets rid of the maintenance overhead of finding unused screen space for new elements.

fishpepper commented:
DanNixon: Please have a look at it again. I added an additional gtest for the positioning.
The RelToAbs conversion should be reworked, looks a bit ugly as it is right now.

Why did you use highX = UNITTEST_DISPLAYPORT_COLS - 2 in your unittest?
I changed it to -1 and it works fine with the updated osd.c code.

I hope my max7456 based fc arrives soon so that I can test on my own...

DanNixon commented:
@fishpepper It appears to be working well now, however the issue about always having a PAL preview still exists.

Logically what happens is correct, however a user would definitely expect the preview to match the actual OSD.

DanNixon commented:
@fishpepper If you are not going to have time to revisit this are you OK with me taking a look at getting this into a mergable state?

mikeller commented:
@fishpepper: Is this still relevant?

mikeller commented:
Closing as stale.
